subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c

```diff
@@ -0,0 +1,761 @@
+#define UPD_STATE_INITIAL 0
+#define UPD_STATE_DOWNLOAD_STARTED 1
+#define UPD_STATE_DOWNLOADED 2
+#define UPD_STATE_DELIVERED 3
+#define UPD_STATE_INSTALLED 4
+
+#define EVENT_PKG_URI_WRITE 0
+#define EVENT_PKG_WRITTEN 1
+#define EVENT_PKG_INTEGRITY_VERIFIED 2
+#define EVENT_INSTALL 4
+#define EVENT_INSTALL_SUCCESSFUL 5
+#define EVENT_INSTALL_FAIL 6
+#define EVENT_DELETE_PACKAGE 7
+#define EVENT_FOR_UPDATE 8
+#define EVENT_DOWNLOAD_FAILED 9
+#define EVENT_PKG_INTEGRITY_FAILED 10
+#define EVENT_ACTIVATE 11
+#define EVENT_DEACTIVATE 12
+
+/*
+ * 0: Initial value. Prior to download any new package in the Device, Update
+ *    Result MUST be reset to this initial value. One side effect of executing
+ *    the Uninstall resource is to reset Update Result to this initial value 0.
+ */
+#define UPD_RES_INITIAL 0
+
+/* 1: Downloading. The package downloading process is ongoing. */
+#define UPD_RES_DOWNLOADING 1
+
+/* 2: Software successfully installed. */
+#define UPD_RES_SW_SUCCESSFULLY_INSTALLED 2
+
+/* 3: Successfully Downloaded and package integrity verified */
+#define UPD_RES_DOWNLOADED_AND_VERIFIED 3
+
+/* 4-49: reserved for expansion of other scenarios>> */
+/* 50: Not enough storage for the new software package. */
+#define UPD_RES_NOT_ENOUGH_STORAGE 50
+
+/* 51: Out of memory during downloading process. */
+#define UPD_RES_OUT_OF_MEMORY_DURING_DOWNLOAD 51
+
+/* 52: Connection lost during downloading process. */
+#define UPD_RES_LOST_CONNECTION_DURING_DOWNLOAD 52
+
+/* 53: Package integrity check failure. */
+#define UPD_RES_PACKAGE_INTEGRITY_CHECK_FAILURE 53
+
+/* 54: Unsupported package type. */
+#define UPD_RES_UNSUPPORTED_PACKAGE_TYPE 54
+
+/* 55: Undefined */
+/* 56: Invalid URI */
+#define UPD_RES_INVALID_URI 56
+
+/* 57: Device defined update error */
+#define UPD_RES_DEVICE_DEFINED_UPDATE_ERROR 57
+
+/* 58: Software installation failure */
+#define UPD_RES_SW_INSTALLATION_FAILURE 58
+
+/* 59: Uninstallation Failure during forUpdate(arg=0) */
+#define UPD_RES_UNINSTALLATION_FAILURE_FOR_UPDATE 59
+
+/*
+ * 60-200 : reserved for expansion selection to be in blocks depending on new
+ *          introduction of features
+ */
+
+struct lwm2m_swmgmt_data {
+	uint16_t obj_inst_id;
+
+	char package_name[PACKAGE_NAME_LEN];
+	char package_version[PACKAGE_VERSION_LEN];
+
+	bool next_package_is_upgrade;
+
+	uint8_t update_state;
+	uint8_t update_result;
+
+	bool activation_state;
+
+	lwm2m_engine_get_data_cb_t read_package_cb;
+	lwm2m_engine_execute_cb_t install_package_cb;
+	lwm2m_engine_user_cb_t upgrade_package_cb;
+	lwm2m_engine_execute_cb_t delete_package_cb;
+	lwm2m_engine_execute_cb_t activate_cb;
+	lwm2m_engine_execute_cb_t deactivate_cb;
+	lwm2m_engine_set_data_cb_t write_package_cb;
+
+#ifdef CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
+	char package_uri[PACKAGE_URI_LEN];
+#endif
+};
+
+/* Package pull request should come with a verify_cb which needs to be stored for when package
+ * gets downloaded
+ */
+static int (*verify_package)(void);
```

Comment:
This doesn't seem to be used at all (only set to NULL in one place), do we need it? I'd rather avoid introducing dead code, it might be confusing for folks analyzing the code.

Reply:
It is called on line 582. However, I believe that callback should be part of `struct lwm2m_swmgmt_data`.

Reply:
Well, I'd rather say the pointer is used, not called - it won't be called as it's only set to NULL in one place. And there's no API to change it value. But I agree with the latter, if it stays it should be part of the struct.

Reply:
I'm not sure about this. Why do we want to store a verify callback per object instance? What's the benefit there?
When I was porting object 9, the idea was to hold a verify callback per pull request. Mostly because this seems to do the job and is cheap memory wise. The API for that is the `struct requesting_object` - that's where one could set the `verify_cb`. Currently it's not really a user facing API but it conveys the idea of not storing a verification callback persistently, per object instance.

Reply:
My thinking is that the verification might differ depending on the object requesting the pull. Each object might have a different way of checking the integrity, depending on the contents of the package.